core/src/main/java/hudson/model/AbstractBuild.java

```diff
@@ -742,13 +742,19 @@ protected final boolean performAllBuildSteps(BuildListener listener, Iterable<?
         }
 
         private void reportError(BuildStep bs, Throwable e, BuildListener listener, boolean phase) {
-            final String publisher = ((Publisher) bs).getDescriptor().getDisplayName();
+            final String buildStep;
+
+            if (bs instanceof Describable) {
+                buildStep = ((Describable) bs).getDescriptor().getDisplayName();
+            } else {
+                buildStep = bs.getClass().getName();
+            }
 
             if (e instanceof AbortException) {
-                LOGGER.log(Level.FINE, "{0} : {1} failed", new Object[] {AbstractBuild.this, publisher});
-                listener.error("Publisher '" + publisher + "' failed: " + e.getMessage());
+                LOGGER.log(Level.FINE, "{0} : {1} failed", new Object[] {AbstractBuild.this, buildStep});
+                listener.error("Step '" + buildStep + "' failed: " + e.getMessage());
```

Comment:
BTW this is fine AFAICT but beware that ASCII `'` is a metacharacter for `Logger.log(Level, String, Object[])`; needs to be escaped as `''`, or just use `‘` and `’`.